runtime/src/iree/base/loop_emscripten.js

```diff
@@ -102,22 +241,87 @@ const IreeLibraryLoopEmscripten = {
         const scope = this.scopes[scope_handle];
         return scope.command_call(callback, user_data, loop);
       }
+
+      iree_loop_command_wait_until(
+          scope_handle, callback, user_data, timeout_ms, loop) {
+        if (!(scope_handle in this.scopes)) return IREE_STATUS_OUT_OF_RANGE;
+
+        const scope = this.scopes[scope_handle];
+        return scope.command_wait_until(
+            callback, user_data, timeout_ms, wait_promise, loop);
+      }
+
+      iree_loop_command_wait_one(
+          scope_handle, callback, user_data, timeout_ms, promise_handle, loop) {
+        if (!(scope_handle in this.scopes)) return IREE_STATUS_OUT_OF_RANGE;
+
+        const scope = this.scopes[scope_handle];
+        const wait_promise = IreeWaitHandlePromise.getPromise(promise_handle);
+        return scope.command_wait_one(
+            callback, user_data, timeout_ms, wait_promise, loop);
+      }
+
+      iree_loop_command_wait_any(
+          scope_handle, callback, user_data, timeout_ms, promise_handles_count,
+          promise_handles, loop) {
+        if (!(scope_handle in this.scopes)) return IREE_STATUS_OUT_OF_RANGE;
+
+        const scope = this.scopes[scope_handle];
+        const wait_promises = [];
+        for (let i = 0; i < promise_handles_count; ++i) {
+          const promise_handle = getValue(promise_handles + i * 4);
+          wait_promises[i] = IreeWaitHandlePromise.getPromise(promise_handle);
+        }
+        return scope.command_wait_any(
+            callback, user_data, timeout_ms, wait_promises, loop);
+      }
+
+      iree_loop_command_wait_all(
+          scope_handle, callback, user_data, timeout_ms, promise_handles_count,
+          promise_handles, loop) {
+        if (!(scope_handle in this.scopes)) return IREE_STATUS_OUT_OF_RANGE;
```

Comment:
INTERNAL or INVALID_ARGUMENT would be more descriptive here (out of range is like an invalid argument with more semantics, but that the scopes are integers is an implementation detail and this is really indicating that either the scope passed in was invalid or there's been some internal error where a scope has disappeared while there were still outstanding users of it)